📝  (INT-2116):  Update for 3PB Documentation (#13)
fahimmahmoodmir/developers-community: pages/Documents/ConversationalAI/Third-PartyBots/CustomEndpoint/basic-content.md

---
pagename: Basic Content
sitesection: Documents
categoryname: "Conversational AI"
documentname: Third-Party Bots
subfoldername: Custom Endpoint
permalink: third-party-bots-custom-endpoint-basic-content.html
indicator:
---

This section will describe The responses that are expected to be sent by the
[Send Conversation Events](third-party-bots-custom-endpoint-service-implementation.html#send-conversation-events)
endpoint of the Custom Endpoint service. To ensure the validity of the response, brands can use our
[Conversation Tester feature](third-party-bots-conversation-tester.html).
In the case of Structure Content validation, you can also leverage
[this tool](https://livepersoninc.github.io/json-pollock/editor/).

### Sending Rich Content (Structured Content)

Structured Content/Rich Content is supported by the core LivePerson platform. Documentation
for the feature can be found [here](getting-started-with-rich-messaging-introduction.html).

{: .important}

> If Images are sent in Rich content, then their URLs must be added to a whitelist via internal

Reviewer: ```suggestion
If images are sent in Rich content, their URLs must be added to a whitelist via internal
```

---

> pagename: Basic Content

Reviewer 2: We are not restricted to the 3 documents structure I've started creating for other vendors.
I think the API Service Specification section is a good start regarding the content. But maybe move them to a separate document: "Implementing a Custom Endpoint service" or something similar. What do you think?

Author: Yep will add the page

permalink: third-party-bots-custom-endpoint-basic-content.html
indicator:
---

### API Service Specification

The Custom Endpoint vendor of Third Party Bots allows brands to build their own API that generates
Bot responses for incoming consumer messages, and connect this to conversations taking place in the
Conversational Cloud. The basic flow of how a consumer message is sent and bot response is received
on an abstract level can be seen in Figure 2.1

<img class="fancyimage" style="width:800px" src="img/customendpoint/message-flow.png">
Figure 2.1 Depicts how a consumer message is sent to Custom Endpoint and a
response is sent back.

<ol>
<li>Consumer Message (CM) is sent to LivePerson Universal Messaging Service (UMS)</li>
<li>CM that was sent to UMS reaches the Third-Party Bot connector</li>
<li>Third-Party Bot connector sends CM to the Custom Endpoint service (i.e `www.mybotapi.com/api` in Figure 2.1) </li>
<li>Custom Endpoint service sends Bot Response (BR) back to the Third-Party Bot connector 
    which parses and validates it</li>
<li>Third-Party Bot connector sends the bot response to the UMS</li>
<li>UMS sends the bot response to the consumer</li>
</ol>

> #### Custom Endpoint (CE) Service Methods

Reviewer 2: Why starting with the additional mentioning of an abbreviation here? If you want to use it explain it in the introduction overview and use it throughout the documentation. Otherwise leave it out i would say.

Author: will be removed

> The Custom Endpoint vendor has a public Github [Repository](https://github.com/LivePersonInc/third-party-bots-custom-endpoint-reference-service)

Reviewer 2: ```suggestion
The Custom Endpoint vendor has a public [GitHub Repository](https://github.com/LivePersonInc/third-party-bots-custom-endpoint-reference-service)
```

> (a.k.a Third-Party Bots Custom Endpoint Reference Service) which contains endpoints and their

Reviewer 2: ```suggestion
which contains the [OpenAPI Specification](https://swagger.io/specification) any service needs to implement in order to connect to Third-Party Bots via the Custom Endpoint Vendor
```

contracts/interface in the format of a swagger file that needs to be implemented by brands.
Following are the methods defined by the specification:

> - **Get Bot Environments**: This API fetches the list of environments defined for the passed `botId`

Reviewer 2: ```suggestion
- **Get Bot Environments**: This endpoint fetches the list of environments defined for the passed `botId`
```

> - **Get Bot State**: This API sends the bot state if the bot is `online` or not.

Reviewer 2: ```suggestion
- **Get Bot State**: This endpoint is called on test connection. It should response with the bot state and the bot version. Every state besides 'online' will let the test connection fail.
```

> - **Get Conversation Events**: This API sends bot responses on the passed consumer messages.

Reviewer 2: ```suggestion
- **Send Conversation Events**: This API sends bot responses on the passed consumer messages.
```
or POST Conversation events if you want to use the REST commands

- **Create Conversation**: This API creates a conversation instance for the given conversation
  (`convId`) for a given `botId`

> Please refer to the [API Service Specification](https://github.com/LivePersonInc/third-party-bots-custom-endpoint-reference-service)

Reviewer 2: We already mentioned the repo just a few lines above

Author: revised in new page

for detailed information on Request and Response bodies created or expected by the Third-Party Bot connector.

#### Authorization and Authentication

> Third-Party Bot connector uses the official Conversational Cloud [OAuth 2.0](accessing-liveperson-apis.html#oauth-20-app-jwt) (App-JWT)

Reviewer 2: ```suggestion
Third-Party Bot connector uses [App-JWT OAuth 2.0](accessing-liveperson-apis.html#oauth-20-app-jwt) authentication 
```

> Authorization mechanism for a server to server interaction. Third-party bot connector uses the provided

Reviewer 2: ```suggestion
mechanism for a server to server interaction. Third-Party Bots uses the provided
```

> `Client ID` and `Client Secret` of App Installation in vendor configuration to generate bearer token.

Reviewer 2: ```suggestion
`Client ID` and `Client Secret` of an App Installation in the vendor configuration to generate bearer tokens.
```

> More information on the sentinel API can be found [here](connector-api-send-api-authorization-and-authentication.html#get-appjwt).

Reviewer 2: ```suggestion
More information on the Sentinel API can be found [here](connector-api-send-api-authorization-and-authentication.html#get-appjwt).
```

> Third-Party Bot connector sends `Authorization` header with the bearer token to all of the requests

Reviewer 2: ```suggestion
Third-Party Bots sends the bearer token inside the `Authorization` on all requests
```

> to Custom Endpoint service. The brands need to ensure authentication and authorization for their

Reviewer 2: ```suggestion
to the Custom Endpoint service. The brands need to ensure the request is authorised, e.g. if the account the request is coming from is allowed to access the addressed bot
```

Custom Endpoint service endpoints.

### Custom Endpoint (CE) Service Flows

This section will describe some of the common interaction flows. We will highlight which endpoints
that are described in the [service methods](third-party-bots-custom-endpoint-basic-content.html#custom-endpoint-ce-service-methods)

> section will be called by the Third-Party Bot connector.

Reviewer 2: Is our Service called Third-Party Bots or Third-Party Bot connector?

Author: revised in new  page

{: .important}
Please note brands needs to store information/session about the Conversation and Bot on their Custom Endpoint
service so that information can be used in subsequent calls.

- Test Connection
- Consumer Message

#### Test Connection

Test connection flow is executed in two different scenarios:

- When a user while creating a Custom Endpoint Bot uses the feature of Test Connection
- When a user starts a bot and Third-Party Bot connector verifies the connection to the Custom Endpoint service

> In both scenarios Figure 2.2 shows the sequence diagram which endpoint in the Custom Endpoint

Reviewer 2: ```suggestion
In both scenarios Figure 2.2 shows the sequence of interactions with the involved services
```

   endpoint of Custom Endpoint service is called with the bearer token received
   from Sentinel API
5. Response from the [Get Bot State](third-party-bots-custom-endpoint-basic-content.html#custom-endpoint-ce-service-methods)
   endpoint is received by Third-Party Bots Connector/API
6. Response of success or failure is sent back based on the response of the Get Bot State endpoint

#### Consumer Message

The consumer message flow is executed once the Third-Party Bot connector receives a consumer message
notification from Universal Message Service (UMS) of Conversational Cloud. Figure 2.3 shows
the sequence diagram of which endpoints in the Custom Endpoint service are called

{: .important}
More information on the Request body and Responses expected by or received from the Custom Endpoint service are available
in [API Service Specification](https://github.com/LivePersonInc/third-party-bots-custom-endpoint-reference-service)

<img class="fancyimage" src="img/customendpoint/consumer-message-flow.png">
Figure 2.3 showing the happy consumer message flow

1. Consumer Message notification is received by the Third-party Bots Connector from UMS

> 2. If Sentinel bearer does not exist Third-Party Bots Connector/API get Sentinel Bearer from

Reviewer 2: I'm not sure we need to include the details of the authentication on our side in the doc, as long as we describe we are sending a bearer generated with the App Install

Author: I think its better to tell the flow from letting customer ask in future what bearer is and where from it is generated

   Sentinel API by using the credentials provided in vendor configuration

> 3. If the received Consumer Message context does not contain a conversation id or a new conversation

Reviewer 2: Consumer message context is not something that has a meaning from an outside context.

Author: changed the words

                ]
              },
              "title": "Recommend a movie"
            }
          ]
        }
      }
    }
  ],
  "analytics": {
    "intents": [
      {
        "id": "base-structured-content",
        "description": "Structured Content Message",
        "confidenceScore": 1
      }
    ]
  }
}
```

Figure 2.5 Showing an example response of Structure Content Message

### Sending Quick Replies (Structured Content)

{: .important}
**Please note** Quick Replies are only supported in Messaging Conversations.

Quick Replies is a special type of Structured Content. It is a message sent along with predefined answers.
For detailed information on Quick Replies check out the documentation for the specific channel ([Mobile SDK and Web](mobile-sdk-and-web-templates-quick-replies-template.html), [Facebook Messenger](facebook-messenger-templates-quick-replies-template.html),
[Google RCS Business Messaging](google-rcs-business-messaging-templates-quick-replies-template.html)).

An example of a Quick Replies message response along with any optional metadata and analytical
information (intents) can be seen below in Figure 2.6. More information on responses expected

> by Third-Party Bot Connector can be found at [API Service Specification](https://github.com/LivePersonInc/third-party-bots-custom-endpoint-reference-service)

Reviewer 2: Doe we really nee to mention this on every content description?

Author: I think yes if someone came only for the Custom Endpoint documentation then it should be there

Reviewer 2: It's not just on the Custom Endpoint once, it's on each content of the custom endpoint

Author: Sorry I mean if someone comes to only a specific content description. I will keep it like that

{: .important}

> The response is expected to be sent by the [Get Conversation Events](third-party-bots-custom-endpoint-basic-content.html#custom-endpoint-ce-service-methods)

Reviewer 2: Do we really need to mention this on every content description? 
We should make have one more detailed description of the POST Conversation Events endpoint in the API description of this documentation instead of repeating this important block again and again

Author: It is intentionally added as a reminder to readers who will come directly to this part of the documentation in case he/she wasn't going on all the document page by page

Reviewer 2: I understand but an approach like this makes the documentation harder to maintain, we already have this issue with information blocks we've add later to some vendors and to some vendors we didn't.
Also what's the expected way of reading this document? I would say a developer won't just implement one single response for a bot but will first get an understanding of how the whole vendor works

Author: This line is not sure will be changed in the future that often so I will keep it for now

            "type": "quickReplies",
            "itemsPerRow": 8,
            "replies": [
              {
                "type": "button",
                "tooltip": "yes i do",
                "title": "yes",
                "click": {
                  "actions": [{ "type": "publishText", "text": "yep" }],
                  "metadata": [{ "type": "ExternalId", "id": "Yes-1234" }]
                }
              },
              {
                "type": "button",
                "tooltip": "No!",
                "title": "No!",
                "click": {
                  "actions": [{ "type": "publishText", "text": "No!" }],
                  "metadata": [{ "type": "ExternalId", "id": "No-4321" }]
                }
              }
            ]
          },
          "message": "Do you like Bots?"
        },
        "metadata": [{ "id": "1234", "type": "ExternalId" }]
      }
    }
  ],
  "analytics": {
    "intents": [
      {
        "id": "base-quick-replies",
        "description": "Quick Replies Messages",
        "confidenceScore": 1
      }
    ]
  }
}
```

Figure 2.6 Showing an example of a Quick Replies response

### Bot Actions

{: .notice}

> Please note we only support **ONE ACTION** per Custom Endpoint service response

Reviewer 2: Shouldn't we have this on every vendor? Or even better in the general section of the documentation

Author: I would still prefer it to have it on Actions to make it even more descriptive so that it's a kind of constant reminder to whoever is looking at a specific part of the document that he must keep it in mind.

Reviewer 2: See my above comment about maintainability of the documentation.
If you stick with this please add it to all vendors

Author: will add new page
